fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1581 @@
+   const FiniteElement &el = *fes.GetFE(0);
+   const IntegrationRule *ir = rule?rule:&DefaultGetRule(el,el);
+   const int dims = el.GetDim();
+   const int symmDims = (dims * (dims + 1)) / 2; // 1x1: 1, 2x2: 3, 3x3: 6
+   const int nq = ir->GetNPoints();
+   dim = mesh->Dimension();
+   ne = fes.GetNE();
+   dofs1D = el.GetOrder() + 1;
+   quad1D = IntRules.Get(Geometry::SEGMENT, ir->GetOrder()).GetNPoints();
+   const GeometryExtension *geo = GeometryExtension::Get(fes,*ir);
+   maps = DofToQuad::Get(fes, fes, *ir);
+   vec.SetSize(symmDims * nq * ne);
+   const double coeff = static_cast<ConstantCoefficient*>(Q)->constant;
+   PADiffusionAssemble(dim, dofs1D, quad1D, ne, maps->W, geo->J, coeff, vec);
+   delete geo;
+}
+
+#ifdef MFEM_USE_OCCA
+// OCCA PA Diffusion MultAdd 2D kernel
+static void OccaPADiffusionMultAdd2D(const int D1D,
+                                     const int Q1D,
+                                     const int NE,
+                                     const double* B,
+                                     const double* G,
+                                     const double* Bt,
+                                     const double* Gt,
+                                     const double* oper,
+                                     const double* solIn,
+                                     double* solOut)
+{
+   const int NUM_QUAD_2D = Q1D*Q1D;
+
+   GET_OCCA_CONST_MEMORY(B);
+   GET_OCCA_CONST_MEMORY(G);
+   GET_OCCA_CONST_MEMORY(Bt);
+   GET_OCCA_CONST_MEMORY(Gt);
+   GET_OCCA_CONST_MEMORY(oper);
+   GET_OCCA_CONST_MEMORY(solIn);
+   GET_OCCA_MEMORY(solOut);
+
+   NEW_OCCA_PROPERTY(props);
+   SET_OCCA_PROPERTY(props, D1D);
+   SET_OCCA_PROPERTY(props, Q1D);
+   SET_OCCA_PROPERTY(props, NUM_QUAD_2D);
+
+   if (!Device::UsingDevice())
+   {
+      NEW_OCCA_KERNEL(MultAdd2D_CPU, fem, diffusion.okl, props);
+      MultAdd2D_CPU(NE,
+                    o_B, o_G,
+                    o_Bt, o_Gt,
+                    o_oper, o_solIn,
+                    o_solOut);
+   }
+   else
+   {
+      NEW_OCCA_KERNEL(MultAdd2D_GPU, fem, diffusion.okl, props);
+      MultAdd2D_GPU(NE,
+                    o_B, o_G,
+                    o_Bt, o_Gt,
+                    o_oper, o_solIn,
+                    o_solOut);
+   }
+}
+#endif // MFEM_USE_OCCA
+
+
+#define QUAD_2D_ID(X, Y) (X + ((Y) * Q1D))
+#define QUAD_3D_ID(X, Y, Z) (X + ((Y) * Q1D) + ((Z) * Q1D*Q1D))
+
+// PA Diffusion MultAdd 2D kernel
+template<const int D1D,
+         const int Q1D> static
+void PADiffusionMultAssembled2D(const int NE,
+                                const double* b,
+                                const double* g,
+                                const double* bt,
+                                const double* gt,
+                                const double* _op,
+                                const double* _x,
+                                double* _y)
+{
+   const int NQ = Q1D*Q1D;
+   const DeviceMatrix B(b,Q1D,D1D);
```

Comment:
This is nice, I like these better than the indexing macros. Does this compile into similar code, where the sizes are known at compile time when accessing the arrays?

Reply:
Without the debug mode that checks for the ranges, it should be similar to the indexing, yes.

Reply:
The size are not known at compilation time in this case. We were considering adding later fixed sized Tensor and also different Layouts for GPU optimizations.

Reply:
We would have to do a Tensor type looking like `Tensor<Q1D,D1D> B(b)` to have the sizes known at compilation time, I agree that this could result in substantial performance gain.

Reply:
@YohannDudouit, I'm not sure I follow what you are saying: `Q1D` and `D1D` are template parameters - they are known at compile time.

Reply:
@v-dobrev I mean that to have what @jakubcerveny wants, we have to give the sizes as template parameters, not as arguments of the constructor. Does that clarify things?

Reply:
Perhaps with the Okina JIT capability we can have the best of both worlds here sometime later.

Reply:
We could already have fixed sized containers here, and avoid dynamic sizes, JIT would only allow us to compile these "kernels" just in time, rather than compiling some of them ahead of time as it is currently. I think that @camierjs  used this class for convenience, because it was already written in one of my branch.
Since I use some template meta-programming to unroll the array index computation (instead of using a `for` loop), maybe that the compiler is capable to inline these sizes already by seeing that the sizes are known at compilation time in this case, but I doubt it.

Reply:
I see what you mean - even though `Q1D` and `D1D` are known at compile time, they are used through the objects where they become dynamic. This can be easily fixed by using the fixed-size `TMatrix` and `TTensor3` classes which I think should work in cuda as well.

Reply:
The question is, how much perfromance can we gain by passing the sizes as template parameters, so that they are known at compile time. Right now, we are probably losing the compile time sizes in the DeviceTensor classes. Are we seeing a performance drop compared to the indexing macros? If not, we might as well stop templating the kernel, everything will be simpler (no need for the unordered_map dispatch) *and* it will work for all orders...

Reply:
I agree we should look at the performance drop. There is also the fact that some of the kernels use local arrays which sizes have to be known at compile time too. The `okina-generic` branch allow to use some memory provided to the kernel for this, but this PR is targeting 4.1 right now.

Reply:
I'm of the opinion that we shoudn't dive down too many rabbit holes for optimizing yet and save more of that for 4.1.  We haven't profiled any of these changes and we have various metaprogramming flavors still in play.

Reply:
I would add that this optimization problem is far from being trivial, the data layout could certainly be optimized for AVX or GPU, and this would certainly have more impact than using the size at compilation time. I agree with @acfisher , the fine optimization of kernels will come later.